modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/resources/META-INF/resources/js/hooks/useStorage.es.js

```diff
@@ -0,0 +1,49 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {useEffect} from 'react';
+
+const defaultDataStorage = {
+	components: new Map(),
+	refs: 0,
+};
+
+const DDMFormRenderer = Symbol('ddm.form.internal');
+
+/**
+ * Simple implementation to create a global context that can be shared
+ * independently of the React tree, Form Renderer is not an application
+ * and can be reused more than once on the same page, so we need to share
+ * requests and fields, they are loaded on demand, so as not to make
```

Comment:
Doesn't the AMD loader already takes care of caching modules? Or is this to prevent making duplicated requests to  retrieve the field types JSON?

Reply:
Yeah, this prevented many requests being made to get the field types and we also saved the component with a wrapper with `React.Lazy` and other details that are in `Field.es.js`. This is because we can have two instances of the Form on the same page and in different trees. Field is also loaded into Rules.